### PR TITLE
Add Javadoc to Asciidoc converter

### PR DESCRIPTION
**Pull request by ppkarwasz**
We add a converter between Javadoc and Asciidoc that converts:

 * from a [`DocCommentTree`](https://docs.oracle.com/en/java/javase/17/docs/api/jdk.compiler/com/sun/source/doctree/DocCommentTree.html), provided by the `javac` compiler or `javadoc` tool,
 * to an AsciiDoctorJ [`Document`](https://javadoc.io/static/org.asciidoctor/asciidoctorj/3.0.0-alpha.2/org/asciidoctor/ast/Document.html).

We provide a primitive implementation of the AsciiDoctorJ API that converts the AST back into a text document.

Part of apache/logging-log4j2#1956

### CODE REVIEW

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AbstractAsciidocTreeVisitor.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        return super.visitLink(node, data);
+    }
+
+    @Override
+    public Void visitLiteral(final LiteralTree node, final AsciidocData data) {
+        if (node.getKind() == DocTree.Kind.CODE) {
+            if (!data.getCurrentLine().isEmpty()) {
+                data.append(" ");
+            }
+            data.append("`").append(node.getBody().getBody()).append("`");
```
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AbstractAsciidocTreeVisitor.java:216]: I think the following cases are not handled:
```
<code>foo`</code>      (needs to be represented with `++foo`++`)
<code>foo</code>bar    (converting this to `foo`bar won't work, it needs to be ``foo``bar)
```

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AsciidocConverter.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import com.sun.source.doctree.DocTree;
+import com.sun.source.doctree.DocTreeVisitor;
+import com.sun.source.doctree.ParamTree;
+import com.sun.source.util.DocTrees;
+import javax.lang.model.element.Element;
+
+/**
+ * Converts a {@link DocCommentTree} into AsciiDoc text.
+ */
+class AsciidocConverter {
```
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AsciidocConverter.java:29]:
```suggestion
final class AsciidocConverter {
```

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AsciidocConverter.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        return data.getDocument().convert();
+    }
+
+    public String toAsciiDoc(final ParamTree tree) {
+        final AsciidocData data = new AsciidocData();
+        tree.accept(PARAM_TREE_VISITOR, data);
+        return data.getDocument().convert();
+    }
+
+    private static class DocCommentTreeVisitor extends AbstractAsciidocTreeVisitor {
```
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AsciidocConverter.java:57]:
```suggestion
    private static final class DocCommentTreeVisitor extends AbstractAsciidocTreeVisitor {
```

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AsciidocConverter.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            for (final DocTree docTree : node.getBody()) {
+                docTree.accept(this, data);
+            }
+            // Flushes the last paragraph
+            data.newParagraph();
+            return super.visitDocComment(node, data);
+        }
+    }
+
+    private static class ParamTreeVisitor extends AbstractAsciidocTreeVisitor {
```
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AsciidocConverter.java:75]:
```suggestion
    private static final class ParamTreeVisitor extends AbstractAsciidocTreeVisitor {
```

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AsciidocData.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+import java.util.EmptyStackException;
+import java.util.function.Function;
+import org.apache.logging.log4j.docgen.processor.internal.BlockImpl;
+import org.apache.logging.log4j.docgen.processor.internal.DocumentImpl;
+import org.apache.logging.log4j.docgen.processor.internal.SectionImpl;
+import org.asciidoctor.ast.Block;
+import org.asciidoctor.ast.Document;
+import org.asciidoctor.ast.StructuralNode;
+
+class AsciidocData {
```
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AsciidocData.java:28]:
```suggestion
final class AsciidocData {
```

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AsciidocData.java
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AsciidocData.java:1]: Shouldn't all the `public` methods be package-local instead?

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/ContentNodeImpl.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+ * limitations under the License.
+ */
+package org.apache.logging.log4j.docgen.processor.internal;
+
+import java.util.List;
+import java.util.Map;
+import org.asciidoctor.ast.ContentNode;
+import org.asciidoctor.ast.Document;
+
+public abstract class ContentNodeImpl implements ContentNode {
```
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/ContentNodeImpl.java:24]:
```suggestion
abstract class ContentNodeImpl implements ContentNode {
```

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/DocumentImpl.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.util.StringBuilderFormattable;
+import org.asciidoctor.ast.Author;
+import org.asciidoctor.ast.Catalog;
+import org.asciidoctor.ast.Document;
+import org.asciidoctor.ast.RevisionInfo;
+import org.asciidoctor.ast.StructuralNode;
+import org.asciidoctor.ast.Title;
+
+public class DocumentImpl extends StructuralNodeImpl implements Document {
```
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/DocumentImpl.java:30]:
```suggestion
public final class DocumentImpl extends StructuralNodeImpl implements Document {
```

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/DocumentImpl.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            buffer.append("= ").append(getTitle()).append("\n\n");
+        }
+        boolean first = true;
+        for (final StructuralNode node : getBlocks()) {
+            if (!first) {
+                buffer.append('\n');
+            } else {
+                first = false;
+            }
+            if (node instanceof final StringBuilderFormattable formattable) {
```
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/DocumentImpl.java:48]: Mind explaining how can we receive an instance of `StringBuilderFormattable`? I was under the impression we are only dealing with ASTs.

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/ListImpl.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+package org.apache.logging.log4j.docgen.processor.internal;
+
+import java.util.Objects;
+import org.apache.logging.log4j.util.StringBuilderFormattable;
+import org.asciidoctor.ast.ContentNode;
+import org.asciidoctor.ast.List;
+import org.asciidoctor.ast.ListItem;
+import org.asciidoctor.ast.StructuralNode;
+
+public class ListImpl extends StructuralNodeImpl implements List {
```
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/ListImpl.java:26]:
```suggestion
public final class ListImpl extends StructuralNodeImpl implements List {
```

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/ListItemImpl.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+ * limitations under the License.
+ */
+package org.apache.logging.log4j.docgen.processor.internal;
+
+import org.apache.logging.log4j.util.StringBuilderFormattable;
+import org.asciidoctor.ast.ContentNode;
+import org.asciidoctor.ast.ListItem;
+import org.asciidoctor.ast.StructuralNode;
+
+public class ListItemImpl extends StructuralNodeImpl implements ListItem {
```
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/ListItemImpl.java:24]:
```suggestion
public final class ListItemImpl extends StructuralNodeImpl implements ListItem {
```

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/RowImpl.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+ * limitations under the License.
+ */
+package org.apache.logging.log4j.docgen.processor.internal;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.asciidoctor.ast.Cell;
+import org.asciidoctor.ast.Row;
+
+public class RowImpl implements Row {
```
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/RowImpl.java:24]:
```suggestion
public final class RowImpl implements Row {
```

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/SectionImpl.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.logging.log4j.util.StringBuilderFormattable;
+import org.asciidoctor.ast.ContentNode;
+import org.asciidoctor.ast.Document;
+import org.asciidoctor.ast.Section;
+import org.asciidoctor.ast.StructuralNode;
+
+public class SectionImpl extends StructuralNodeImpl implements Section {
```
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/SectionImpl.java:27]:
```suggestion
public final class SectionImpl extends StructuralNodeImpl implements Section {
```

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/TableImpl.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.logging.log4j.util.StringBuilderFormattable;
+import org.asciidoctor.ast.Column;
+import org.asciidoctor.ast.ContentNode;
+import org.asciidoctor.ast.Row;
+import org.asciidoctor.ast.Table;
+
+public class TableImpl extends StructuralNodeImpl implements Table {
```
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/TableImpl.java:27]:
```suggestion
public final class TableImpl extends StructuralNodeImpl implements Table {
```

File: log4j-docgen/src/test/it/example/JavadocExample.java
Comment (vy) [inline log4j-docgen/src/test/it/example/JavadocExample.java:1]: Could you add tests for the following too, please?
```
<code>foo</code>
<code>foo`</code>
<code>`foo</code>
<code>`foo`</code>
<code>foo</code>bar
<code>foo`</code>bar
<code>`foo</code>bar
<code>`foo`</code>bar
<code>foo``</code>
<code>``foo</code>
<code>``foo``</code>
<code>foo</code>bar
<code>foo``</code>bar
<code>``foo</code>bar
<code>``foo``</code>bar
{@code foo}
{@code foo`}
{@code `foo}
{@code `foo`}
{@code foo}bar
{@code foo`}bar
{@code `foo}bar
{@code `foo`}bar
{@code foo``}
{@code ``foo}
{@code ``foo``}
{@code foo}bar
{@code foo``}bar
{@code ``foo}bar
{@code ``foo``}bar
<pre>
breaking the `source` block:
----
've just done it
</pre>
<pre>[<![CDATA[
<code>this should not get converted</code>
{@code what about this?}
]]></pre>
<ul>
<li>item without ending
<li>- breaking the item</li>
<li>* breaking the item</li>
<li>. breaking the item</li>
</ul>
```
Note that we don't need to accept every broken or non-idiomatic javadoc. Above are just certain corner cases I am able to think of. We can also choose to rewrite the source javadoc causing the issue.

File: log4j-docgen/src/test/resources/expected/processor/JavadocExample.adoc [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+| B
+| 2
+
+|===
+
+=== Subsection
+
+[source]
+----
+     private static final Logger logger = LogManager.getLogger();
```
Comment (vy) [inline log4j-docgen/src/test/resources/expected/processor/JavadocExample.adoc:104]: Can we remove the leading whitespace?

**Review by vy — CHANGES_REQUESTED** [review]

File: log4j-docgen/pom.xml [hunk 15]
```diff
@@ -27,6 +27,7 @@
   <artifactId>log4j-docgen</artifactId>
 
   <properties>
+    <maven.compiler.release>17</maven.compiler.release>
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-docgen/pom.xml:30]: The `log4j-plugins` artifact is a **required** dependency (since I am actually **using** the annotation interfaces in the code), so the bytecode levels must match.

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AsciidocData.java
Comment (ppkarwasz) [reply log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AsciidocData.java:1]: I guess it is a matter of style, they are protected anyway.

I thought about moving the internals of the `AsciidocConverter` to an `internal` package. If the methods remain `public`, this can be done without a single change.

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/DocumentImpl.java (hunk 8 above)
Comment (ppkarwasz) [reply log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/internal/DocumentImpl.java:48]: Premature optimalization? I added `StringBuilderFormattable` to all AST node implementations.

File: log4j-docgen/src/test/resources/expected/processor/JavadocExample.adoc (hunk 14 above)
Comment (ppkarwasz) [reply log4j-docgen/src/test/resources/expected/processor/JavadocExample.adoc:104]: That line is actually in a `<pre>` tag? Should we strip leading spaces there?

I could strip the same amount of leading whitespace from each row.

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AbstractAsciidocTreeVisitor.java (hunk 1 above)
Comment (ppkarwasz) [reply log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/AbstractAsciidocTreeVisitor.java]: I didn't implement the second one, since it requires some sort of lookahead logic, but I transformed `<code>foo</code>bar` into `` `foo` bar``.

File: log4j-docgen/src/test/resources/expected/processor/JavadocExample.adoc (hunk 14 above)
Comment (ppkarwasz) [reply log4j-docgen/src/test/resources/expected/processor/JavadocExample.adoc]: Done: the leading space common to all the lines in the `<pre>` block is removed. Only relative indentation is left.
